// leetcode-query/lib/index.js
"use strict";var D=Object.create;var v=Object.defineProperty;var $=Object.getOwnPropertyDescriptor;var O=Object.getOwnPropertyNames;var F=Object.getPrototypeOf,z=Object.prototype.hasOwnProperty;var G=(n,t)=>{for(var e in t)v(n,e,{get:t[e],enumerable:!0})},U=(n,t,e,i)=>{if(t&&typeof t=="object"||typeof t=="function")for(let s of O(t))!z.call(n,s)&&s!==e&&v(n,s,{get:()=>t[s],enumerable:!(i=$(t,s))||i.enumerable});return n};var p=(n,t,e)=>(e=n!=null?D(F(n)):{},U(t||!n||!n.__esModule?v(e,"default",{value:n,enumerable:!0}):e,n)),M=n=>U(v({},"__esModule",{value:!0}),n);var H={};G(H,{BASE_URL:()=>m,BASE_URL_CN:()=>f,Cache:()=>S,Credential:()=>k,LeetCode:()=>_,LeetCodeCN:()=>A,Mutex:()=>y,RateLimiter:()=>g,USER_AGENT:()=>c,cache:()=>h,caches:()=>B,default:()=>V});module.exports=M(H);var N=p(require("events")),P=p(require("node-fetch"));var S=class{constructor(){this._table={}}get(t){let e=this._table[t];if(e){if(e.expires>Date.now())return e.value;this.remove(t)}return null}set(t,e,i=6e4){this._table[t]={key:t,value:e,expires:i>0?Date.now()+i:0}}remove(t){delete this._table[t]}clear(){this._table={}}load(t){this._table=JSON.parse(t)}},h=new S,B={default:h};var m="https://leetcode.com",f="https://leetcode-cn.com",c="Mozilla/5.0 LeetCode API";var q=p(require("node-fetch"));function d(n){return n.split(";").map(t=>t.trim().split("=")).reduce((t,e)=>(t[e[0]]=e[1],t),{})}async function j(){let n=await(0,q.default)(m,{headers:{"user-agent":c}}).then(e=>e.headers.get("set-cookie"));return d(n).csrftoken}var k=class{constructor(t){t&&(this.session=t.session,this.csrf=t.csrf)}async init(t){return this.csrf=await j(),t&&(this.session=t),this}};var I=p(require("events")),y=class extends I.default{constructor(e=1){super();this.space=e,this.used=0,this.releases=[]}async lock(){if(this.used>=this.space){let e=new Promise(i=>this.releases.push(i));this.emit("wait",{lock:e,release:this.releases[this.releases.length-1]}),await e}return this.used++,this.emit("lock"),this.used}unlock(){var e;return this.used<=0?0:(this.releases.length>0&&((e=this.releases.shift())==null||e()),this.used--,this.emit("unlock"),this.used<=0&&this.emit("all-clear"),this.used)}resize(e){var i;for(this.space=e;this.used<e&&this.releases.length>0;)(i=this.releases.shift())==null||i();return this.space}full(){return this.used>=this.space}waiting(){return this.releases.length}},g=class extends y{constructor({limit:e=20,interval:i=1e4,concurrent:s=2}={}){super(s);this.count=0;this.last=0;this.time_mutex=new y(e),this.interval=i,this.time_mutex.on("lock",(...o)=>this.emit("time-lock",...o)),this.time_mutex.on("unlock",(...o)=>this.emit("time-unlock",...o))}async lock(){return this.last+this.interval<Date.now()?this.reset():this.time_mutex.full()&&!this.timer&&this.cleaner(),await this.time_mutex.lock(),this.count++,super.lock()}reset(){for(;this.count>0;)this.time_mutex.unlock(),this.count--;this.last=Date.now(),this.emit("timer-reset")}cleaner(){this.timer=setTimeout(()=>{this.reset(),setTimeout(()=>{this.time_mutex.waiting()>0?this.cleaner():this.timer=void 0},0)},this.last+this.interval-Date.now())}set limit(e){this.time_mutex.resize(e)}};var R=class extends N.default{constructor(e=null,i=h){super();this.limiter=new g;let s;this.initialized=new Promise(o=>{s=o}),this.cache=i,e?(this.credential=e,setImmediate(()=>s())):(this.credential=new k,this.credential.init().then(()=>s()))}async user(e){await this.initialized;let{data:i}=await this.graphql({operationName:"getUserProfile",variables:{username:e},query:`
            query getUserProfile($username: String!) {
                allQuestionsCount {
                    difficulty
                    count
                }
                matchedUser(username: $username) {
                    username
                    socialAccounts
                    githubUrl
                    contributions {
                        points
                        questionCount
                        testcaseCount
                    }
                    profile {
                        realName
                        websites
                        countryName
                        skillTags
                        company
                        school
                        starRating
                        aboutMe
                        userAvatar
                        reputation
                        ranking
                    }
                    submissionCalendar
                    submitStats {
                        acSubmissionNum {
                            difficulty
                            count
                            submissions
                        }
                        totalSubmissionNum {
                            difficulty
                            count
                            submissions
                        }
                    }
                    badges {
                        id
                        displayName
                        icon
                        creationDate
                    }
                    upcomingBadges {
                        name
                        icon
                    }
                    activeBadge {
                        id
                    }
                }
                recentSubmissionList(username: $username, limit: 20) {
                    title
                    titleSlug
                    timestamp
                    statusDisplay
                    lang
                }
            }
            `});return i}async user_contest_info(e){await this.initialized;let{data:i}=await this.graphql({operationName:"userContestRankingInfo",variables:{username:e},query:`
            query userContestRankingInfo($username: String!) {
                userContestRanking(username: $username) {
                    attendedContestsCount
                    rating
                    globalRanking
                    totalParticipants
                    topPercentage
                    badge {
                        name
                    }
                }
                userContestRankingHistory(username: $username) {
                    attended
                    trendDirection
                    problemsSolved
                    totalProblems
                    finishTimeInSeconds
                    rating
                    ranking
                    contest {
                        title
                        startTime
                    }
                }
            }
            `});return i}async recent_submissions(e,i=20){await this.initialized;let{data:s}=await this.graphql({operationName:"getRecentSubmissionList",variables:{username:e,limit:i},query:`query getRecentSubmissionList($username: String!, $limit: Int) {
                recentSubmissionList(username: $username, limit: $limit) {
                    title
                    titleSlug
                    timestamp
                    statusDisplay
                    lang
                }
            }`});return s.recentSubmissionList||[]}async submissions({limit:e=20,offset:i=0,slug:s}={}){await this.initialized;let o=[],b=new Set,a=i;for(;o.length<e;){let{data:l}=await this.graphql({operationName:"Submissions",variables:{offset:a,limit:e-o.length>20?20:e-o.length,slug:s},query:`query Submissions($offset: Int!, $limit: Int!, $slug: String) {
                    submissionList(offset: $offset, limit: $limit, questionSlug: $slug) {
                        hasNext
                        submissions { id lang time timestamp statusDisplay runtime url isPending title memory titleSlug }
                    }
                }`});for(let r of l.submissionList.submissions)r.id=parseInt(r.id,10),r.timestamp=parseInt(r.timestamp,10)*1e3,r.isPending=r.isPending!=="Not Pending",r.runtime=parseInt(r.runtime,10)||0,r.memory=parseFloat(r.memory)||0,!b.has(r.id)&&(b.add(r.id),o.push(r));if(!l.submissionList.hasNext)break;a+=20}return o}async submission(e){var i;await this.initialized;try{await this.limiter.lock();let b=(i=(await(await(0,P.default)(`${m}/submissions/detail/${e}/`,{headers:{origin:m,referer:m,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"user-agent":c}})).text()).match(/var pageData = ({[^]+?});/))==null?void 0:i[1],a=new Function("return "+b)(),l={id:parseInt(a.submissionId),problem_id:parseInt(a.questionId),runtime:parseInt(a.runtime),runtime_distribution:a.runtimeDistributionFormatted?JSON.parse(a.runtimeDistributionFormatted).distribution.map(r=>[+r[0],r[1]]):[],runtime_percentile:0,memory:parseInt(a.memory),memory_distribution:a.memoryDistributionFormatted?JSON.parse(a.memoryDistributionFormatted).distribution.map(r=>[+r[0],r[1]]):[],memory_percentile:0,code:a.submissionCode,details:a.submissionData};return l.runtime_percentile=l.runtime_distribution.reduce((r,[w,x])=>r+(w>=l.runtime?x:0),0),l.memory_percentile=l.memory_distribution.reduce((r,[w,x])=>r+(w>=l.memory/1e3?x:0),0),this.limiter.unlock(),l}catch(s){throw this.limiter.unlock(),s}}async problems({category:e="",offset:i=0,limit:s=100,filters:o={}}={}){await this.initialized;let b={categorySlug:e,skip:i,limit:s,filters:o},{data:a}=await this.graphql({operationName:"problemsetQuestionList",variables:b,query:`query problemsetQuestionList($categorySlug: String, $limit: Int, $skip: Int, $filters: QuestionListFilterInput) {
                problemsetQuestionList: questionList( 
                    categorySlug: $categorySlug
                    limit: $limit
                    skip: $skip
                    filters: $filters
                ) {
                    total: totalNum
                    questions: data {
                        acRate 
                        difficulty 
                        freqBar 
                        questionFrontendId 
                        isFavor
                        isPaidOnly 
                        status 
                        title 
                        titleSlug 
                        topicTags { name id slug } 
                        hasSolution 
                        hasVideoSolution
                    }
                }
            }`});return a.problemsetQuestionList}async problem(e){await this.initialized;let{data:i}=await this.graphql({operationName:"questionData",variables:{titleSlug:e.toLowerCase().replace(/\s/g,"-")},query:`query questionData($titleSlug: String!) {
                question(titleSlug: $titleSlug) {
                  questionId
                  questionFrontendId
                  boundTopicId
                  title
                  titleSlug
                  content
                  translatedTitle
                  translatedContent
                  isPaidOnly
                  difficulty
                  likes
                  dislikes
                  isLiked
                  similarQuestions
                  exampleTestcases
                  contributors {
                    username
                    profileUrl
                    avatarUrl
                  }
                  topicTags {
                    name
                    slug
                    translatedName
                  }
                  companyTagStats
                  codeSnippets {
                    lang
                    langSlug
                    code
                  }
                  stats
                  hints
                  solution {
                    id
                    canSeeDetail
                    paidOnly
                    hasVideoSolution
                    paidOnlyVideo
                  }
                  status
                  sampleTestCase
                  metaData
                  judgerAvailable
                  judgeType
                  mysqlSchemas
                  enableRunCode
                  enableTestMode
                  enableDebugger
                  envInfo
                  libraryUrl
                  adminUrl
                  challengeQuestion {
                    id
                    date
                    incompleteChallengeCount
                    streakCount
                    type
                  }
                  note
                }
              }`});return i.question}async whoami(){await this.initialized;let{data:e}=await this.graphql({operationName:"globalData",variables:{},query:`query globalData {
                userStatus {
                  userId
                  username
                  avatar
                  isSignedIn
                  isMockUser
                  isPremium
                  isAdmin
                  isSuperuser
                  isTranslator
                  permissions
                }
              }`});return e.userStatus}async graphql(e){await this.initialized;try{await this.limiter.lock();let i=m,s=await(0,P.default)(`${i}/graphql`,{method:"POST",headers:{"content-type":"application/json",origin:i,referer:i,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"x-csrftoken":this.credential.csrf||"","user-agent":c},body:JSON.stringify(e)});if(this.emit("receive-graphql",s.clone()),s.headers.has("set-cookie")){let o=d(s.headers.get("set-cookie"));o.csrftoken&&(this.credential.csrf=o.csrftoken,this.emit("update-csrf",this.credential))}return this.limiter.unlock(),s.json()}catch(i){throw this.limiter.unlock(),i}}},_=R;var Q=p(require("events")),E=p(require("node-fetch"));var T=p(require("node-fetch"));async function J(){let n=await(0,T.default)(f,{headers:{"user-agent":c}}).then(e=>e.headers.get("set-cookie"));return d(n).csrftoken}var C=class{constructor(t){t&&(this.session=t.session,this.csrf=t.csrf)}async init(t){return this.csrf=await J(),t&&(this.session=t),this}};var L=class extends Q.default{constructor(e=null,i=h){super();this.limiter=new g;let s;this.initialized=new Promise(o=>{s=o}),this.cache=i,e?(this.credential=e,setImmediate(()=>s())):(this.credential=new C,this.credential.init().then(()=>s()))}async user(e){await this.initialized;let{data:i}=await this.graphql({operationName:"getUserProfile",variables:{username:e},query:`
            query getUserProfile($username: String!) {
                userProfileUserQuestionProgress(userSlug: $username) {
                    numAcceptedQuestions { difficulty count }
                    numFailedQuestions { difficulty count }
                    numUntouchedQuestions { difficulty count }
                }
                userProfilePublicProfile(userSlug: $username) {
                    username haveFollowed siteRanking
                    profile { 
                        userSlug realName aboutMe userAvatar location gender websites skillTags contestCount asciiCode
                        medals { name year month category }
                        ranking { 
                            currentLocalRanking currentGlobalRanking currentRating totalLocalUsers totalGlobalUsers
                        }
                        socialAccounts { provider profileUrl }
                    }
                }
            }
            `});return i}async graphql(e){await this.initialized;try{await this.limiter.lock();let i=f,s=await(0,E.default)(`${i}/graphql`,{method:"POST",headers:{"content-type":"application/json",origin:i,referer:i,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"x-csrftoken":this.credential.csrf||"","user-agent":c},body:JSON.stringify(e)});if(this.emit("receive-graphql",s.clone()),s.headers.has("set-cookie")){let o=d(s.headers.get("set-cookie"));o.csrftoken&&(this.credential.csrf=o.csrftoken,this.emit("update-csrf",this.credential))}return this.limiter.unlock(),s.json()}catch(i){throw this.limiter.unlock(),i}}},A=L;var V=_;0&&(module.exports={BASE_URL,BASE_URL_CN,Cache,Credential,LeetCode,LeetCodeCN,Mutex,RateLimiter,USER_AGENT,cache,caches});
